packages/wasi/src/index.ts

```diff
@@ -761,6 +891,26 @@ export default class WASIDefault {
           nread: number
         ) => {
           const stats = CHECK_FD(fd, WASI_RIGHT_FD_READ | WASI_RIGHT_FD_SEEK);
+          if (this.asyncified) {
+            return asyncify(async () => {
+              let read = 0;
+              for (const iov of getiovs(iovs, iovsLen)) {
+                let r = 0;
+                while (r < iov.byteLength) {
+                  await promisify(fs.read)(
```

Comment:
I think it would be better to use the `readSync` API rather than make it awaitable and wait for the result.

```suggestion
                  fs.readSync(
```

Reply:
But then it wouldn't be asynchronous? Why would a promise be needed then? I probably should have explained my use case more, I'm looking to run a wasi program in a worker with stdio being passed through postMessage calls. Right now, it's impossible to block while waiting for an stdin message because if I just do something like `while (!haveStdin) {}`, control never gets passed to the js event loop and the message handler never runs, so I'd like to be able to do something like (*very* rough sketch): `fs.read = (...args) => new Promise(resolve => addStdinListener(resolve));`

Reply:
Would that be something that you'd be interested in adding to this library? It would make it easier for me to debug that functionality from inside of the wasi runtime rather than from outside, and I can't think of many other use cases for async i/o in wasi in the browser other than something like this.

Reply:
> But then it wouldn't be asynchronous? Why would a promise be needed then?

Yeah, we dealt with a similar problem in the WebAssembly shell. In general, asyncify resolves that, so I think that `fs.readSync` should just work :) (but I haven't tested it).

If you can test it would be great. If it works, then let's go with `fs.readSync`. If that doesn't work locally for you, then let's use your current solution. Let m eknow.

PS: Yeah, we are interested on adding it to the library!